### ISSUE TITLE
Speed up tile building

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#3389]
I'm not sure if there's already an existing issue about it, but maybe better to start fresh anyways. on a medium-sized machine (e.g. 16 threads) it takes approx 1.5 - 2 days to build the planet. more threads can definitely bring it down to less than 24 hours, but maybe there's a few things we could do in terms of efficiency. please add thoughts, comments or other ideas.

- thread work distribution: e.g. while building the tiles, initially the threads are passed the tile list (or actually list of nodes if I read right) which are sorted by graphid, then by osm id. the first threads are finished rather quickly, then a looong tail of the last threads to finish. since there's already a `tile_manifest.json`, wouldn't it be a low-hanging fruit to extend that logic to sort the tiles by amount of nodes/edges they contain before slicing them up to pass them to the threads? that might kill performance when building tiles (which is serializing the objects sorted by graph id)? but if threads would get a contiguous chunk of the sorted IDs it should work? couldn't quite figure out how it's currently being handled if multiple threads have to build objects in the same tile (or if they don't, how that's being avoided).
- multithread the parsing stage, which will have similar problems around jumping around the id space, though there's already a rough description by @kevinkreiser offline (I can post that if you're ok with it). probably quite a bit more work than the first point (if that's feasible).

**Comment by nilsnolde** [comment]
re my first point. just starting to realize that "low-hanging fruit" is probably very optimistic..

**Comment by kevinkreiser** [comment]
yeah regarding the first optimization. because it relies on iterating through a binary file there are two things to keep in mind:

1. the threads know which range of tiles in the sequence they are about
2. jumping around in the sequence might be expensive but probably not a huge deal

to fix the first one we can change easily, instead of giving them a range we give them a set of tiles and where they are in the sequence. load balancing by edge count is a good idea so that each thread does approximately the same number of edges in total.

the second one we just have to measure and see what happens. it might be that for ssds especially nvme ones there is no impact while spinning rust is crippled by jumping around. i'm not sure about it.


regarding multithreading the parsing stage, sure i dont mind at all if you quote me on the stuff i was spitballing to you. feel free

**Comment by nilsnolde** [comment]
> to fix the first one we can change easily

yeah cool I'd give it a spin soon. from what I can see it's a pretty quick thing to try out.

> while spinning rust is crippled by jumping around. i'm not sure about it

I can definitely test the best case of nvme's, but don't have access to any hdd's. do you?

here more info on multithreading the parsing stuff:
_the hard part is making it work with what valhalla needs to build a graph. i think we can probably do it by adding more sorting steps to the processing. specifically, if im remembering right, essentially we need to write 1 sequence file per thread and then do a merge sort when they are done to ge tthe final sorted sequence. its not too difficult in my opinion but we'll need to care a bunch about which nodes we are keeping and which not, as we dont keep them in a hash map but rather sorted by id. if different threads are jumping around all over the id space this will kill parsing performance. i can think of two ways to handle this, one that requires some synchronization and another that doesnt. i would suggest the later where each thread works on a contiguous range of the planet file and so can keep track of its place in the sorted list of osm nodes_

**Comment by dnesbitt61** [comment]
One low-hanging fruit could be to improve admin processing in the graph build stage. I made a recent large improvement in performance by changing from unordered_multimap to std::multimap (see PR https://github.com/valhalla/valhalla/pull/3427).

Another possibility is to update the admin logic to check if the state admin query only returns 1 mutlipolygon result. If so, and the tile's bounding box is totally inside that multi-polygon then you can return that one admin and then set tile_withing_one_admin within graphbuilder.cc. This would have removed the need to do point in poly for all other nodes within the tile - providing additional performance improvement.

Unfortunately this led to a mismatch in tile sizes - so some admin information must have changed. I did not pursue what the nature of these differences were and whether they indicate an issue or not. In theory this approach seems like it should work and my performance testing showed an additional gain above what the initial PR provided.

**Comment by kinkard** [comment]
Seems no longer an issue, see https://github.com/valhalla/valhalla/issues/5099 and linked PRs for more details
